packages/protocol/contracts/L1/verifiers/SGXVerifier.sol

```diff
@@ -7,38 +7,156 @@
 pragma solidity ^0.8.20;
 
 import { EssentialContract } from "../../common/EssentialContract.sol";
+import { ECDSAUpgradeable } from "@ozu/utils/cryptography/ECDSAUpgradeable.sol";
 import { Proxied } from "../../common/Proxied.sol";
 
 import { TaikoData } from "../TaikoData.sol";
 
 import { IVerifier } from "./IVerifier.sol";
 
 /// @title GuardianVerifier
-// TODO(dani): implement this verifier.
 contract SGXVerifier is EssentialContract, IVerifier {
-    uint256[50] private __gap;
+    event InstanceAdded(
+        uint256 indexed instanceId, address indexed instanceInitPubKey
+    );
 
-    error UNIMPLEMENTED();
+    struct ProofData {
+        uint256 id;
+        address newPubKey;
+        bytes signature;
+    }
+
+    /// @dev For gas savings, we shall assign each SGX instance with an id
+    /// so that when we need to set a new pub key, just write storage once.
+    uint256 public uniqueVerifiers;
+
+    /// @dev One SGX instance is uniquely identified (on-chain) by it's ECDSA
+    /// public key. Once that public key is used (by proof verification) it has
+    /// to be overwritten by a new one (representing the same instance). This is
+    /// due to side-channel protection.
+    mapping(uint256 instanceId => address sgxInstance) public sgxRegistry;
+
+    uint256[48] private __gap;
+
+    error SGX_NOT_VALID_SIGNER_OR_ID_MISMATCH();
+    error SGX_INVALID_PROOF_SIZE();
 
     /// @notice Initializes the contract with the provided address manager.
     /// @param _addressManager The address of the address manager contract.
     function init(address _addressManager) external initializer {
         EssentialContract._init(_addressManager);
     }
 
+    /// @notice Adds trusted SGX instances to the registry.
+    /// @param trustedInstances The address array of trusted SGX instances.
+    function addToRegistryByOwner(address[] memory trustedInstances)
+        external
+        onlyOwner
+    {
+        for (uint256 i; i < trustedInstances.length; i++) {
+            sgxRegistry[uniqueVerifiers] = trustedInstances[i];
+
+            emit InstanceAdded(uniqueVerifiers, trustedInstances[i]);
+
+            uniqueVerifiers++;
+        }
```

Comment:
Can also be extracted to some internal function so it can be reused by `addToRegistryByOwner` and `addToRegistryBySgxInstance`.